File: lib/core/database/daos/task_dao.dart
import 'package:drift/drift.dart';

import '../../task/task_state.dart';
import '../app_database.dart';
import '../tables/tasks.dart';

part 'task_dao.g.dart';

@DriftAccessor(tables: [Tasks])
class TaskDao extends DatabaseAccessor<AppDatabase> with _$TaskDaoMixin {
  TaskDao(super.db);

  Future<List<TaskEntity>> getAll() => select(tasks).get();
  Future<TaskEntity?> getById(String id) =>
      (select(tasks)..where((t) => t.id.equals(id))).getSingleOrNull();

  /// 活跃任务：waiting / running / paused
  Stream<List<TaskEntity>> watchActive() {
    return (select(tasks)
          ..where((t) => t.status.isIn(
              const [TaskStatus.waiting, TaskStatus.running, TaskStatus.paused]))
          ..orderBy([(t) => OrderingTerm.asc(t.createdAt)]))
        .watch();
  }

  Stream<List<TaskEntity>> watchByStatus(List<String> statuses) {
    return (select(tasks)
          ..where((t) => t.status.isIn(statuses))
          ..orderBy([(t) => OrderingTerm.desc(t.updatedAt)]))
        .watch();
  }

  Stream<TaskEntity?> watchById(String id) {
    return (select(tasks)..where((t) => t.id.equals(id))).watchSingleOrNull();
  }

  Future<List<TaskEntity>> getByStatus(List<String> statuses) {
    return (select(tasks)..where((t) => t.status.isIn(statuses))).get();
  }

  Future<String> insertTask(TasksCompanion entry) async {
    final id = entry.id.present ? entry.id.value : attachedDatabase.newId();
    await into(tasks).insert(entry.copyWith(id: Value(id)));
    return id;
  }

  Future<int> updateProgress(String id, int progressBytes, int totalBytes) {
    return (update(tasks)..where((t) => t.id.equals(id))).write(
      TasksCompanion(
        progressBytes: Value(progressBytes),
        totalBytes: Value(totalBytes),
        updatedAt: Value(DateTime.now()),
      ),
    );
  }

  Future<int> updateStatus(
    String id,
    String status, {
    String? errorMessage,
    int? retryCount,
  }) {
    return (update(tasks)..where((t) => t.id.equals(id))).write(
      TasksCompanion(
        status: Value(status),
        errorMessage: Value(errorMessage),
        retryCount: retryCount != null ? Value(retryCount) : const Value.absent(),
        updatedAt: Value(DateTime.now()),
      ),
    );
  }

  Future<int> updatePayloadJson(String id, String payloadJson) {
    return (update(tasks)..where((t) => t.id.equals(id))).write(
      TasksCompanion(
        payloadJson: Value(payloadJson),
        updatedAt: Value(DateTime.now()),
      ),
    );
  }

  Future<int> deleteById(String id) =>
      (delete(tasks)..where((t) => t.id.equals(id))).go();

  /// 启动时调用：把上次未结束的 running 改回 waiting，准备恢复。
  Future<int> reviveInterrupted() {
    return (update(tasks)..where((t) => t.status.equals(TaskStatus.running))).write(
      TasksCompanion(status: const Value(TaskStatus.waiting), updatedAt: Value(DateTime.now())),
    );
  }
}

File: app/src/main/java/com/nastools/app/data/database/dao/TaskDao.kt
package com.nastools.app.data.database.dao

import androidx.room.*
import com.nastools.app.data.database.entity.TaskEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface TaskDao {
    @Query("SELECT * FROM tasks WHERE status IN (:statuses) ORDER BY priority DESC, createdAt ASC")
    fun observeByStatus(statuses: List<String>): Flow<List<TaskEntity>>

    @Query("SELECT * FROM tasks WHERE status IN ('waiting', 'running', 'paused') ORDER BY priority DESC, createdAt ASC")
    fun observeActive(): Flow<List<TaskEntity>>

    @Query("SELECT * FROM tasks WHERE id = :id")
    suspend fun getById(id: String): TaskEntity?

    @Query("SELECT * FROM tasks WHERE id = :id")
    fun observeById(id: String): Flow<TaskEntity?>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(task: TaskEntity)

    @Update
    suspend fun update(task: TaskEntity)

    @Query("UPDATE tasks SET status = :status, updatedAt = :updatedAt WHERE id = :id")
    suspend fun updateStatus(id: String, status: String, updatedAt: Long = System.currentTimeMillis())

    @Query("UPDATE tasks SET status = :status, errorMessage = :errorMessage, retryCount = :retryCount, updatedAt = :updatedAt WHERE id = :id")
    suspend fun updateStatusWithError(
        id: String,
        status: String,
        errorMessage: String?,
        retryCount: Int,
        updatedAt: Long = System.currentTimeMillis()
    )

    @Query("UPDATE tasks SET progressBytes = :progressBytes, totalBytes = :totalBytes, updatedAt = :updatedAt WHERE id = :id")
    suspend fun updateProgress(
        id: String,
        progressBytes: Long,
        totalBytes: Long,
        updatedAt: Long = System.currentTimeMillis()
    )

    @Query("UPDATE tasks SET payloadJson = :payloadJson, updatedAt = :updatedAt WHERE id = :id")
    suspend fun updatePayload(id: String, payloadJson: String, updatedAt: Long = System.currentTimeMillis())

    @Query("UPDATE tasks SET title = :title, updatedAt = :updatedAt WHERE id = :id")
    suspend fun updateTitle(id: String, title: String, updatedAt: Long = System.currentTimeMillis())

    @Query("UPDATE tasks SET status = 'waiting' WHERE status = 'running'")
    suspend fun reviveInterrupted(): Int

    @Query("DELETE FROM tasks WHERE id = :id")
    suspend fun deleteById(id: String)

    @Query("DELETE FROM tasks WHERE id IN (:ids)")
    suspend fun deleteByIds(ids: List<String>)

    @Query("DELETE FROM tasks")
    suspend fun deleteAll()
}
